redbiom/admin.py

```
@@ -0,0 +1,324 @@
+    post = redbiom.requests.make_post(config)
+    get = redbiom.requests.make_get(config)
+
+    redbiom.requests.valid(context, get)
+
+    tab = biom.load_table(table)
+    obs = tab.ids(axis='observation')
+    samples = tab.ids()
+
+    if not redbiom.util.has_sample_metadata(samples):
+        raise ValueError("Sample metadata must be loaded first.")
+
+    represented = get(context, 'SMEMBERS', 'samples-represented-data')
+    if set(samples).intersection(set(represented)):
+        raise ValueError("At least one sample to load already exists")
+
+    acquired = False
+    while not acquired:
+        # not using redlock as time interval isn't that critical
+        acquired = get(context, 'SETNX', '__load_table_lock/1') == 1
+        if not acquired:
+            # TODO: support verbose
+            time.sleep(1)
+
+    try:
+        # load the observation index
+        obs_index = get(context, 'GET', '__observation_index')
+        if obs_index is None:
+            obs_index = {}
+        else:
+            obs_index = json.loads(obs_index)
+
+        # update the observation index if and where necessary
+        new_ids = {i for i in obs if i not in obs_index}
+        id_start = max(obs_index.values() or [-1]) + 1
+        for idx, id_ in zip(range(id_start, id_start + len(new_ids)), new_ids):
+            obs_index[id_] = idx
+
+        # load up the table per-sample
+        for values, id_, _ in tab.iter(dense=False):
+            int_values = values.astype(int)
+            remapped = [obs_index[i] for i in obs[values.indices]]
+
+            packed = '\t'.join(["%s\t%d" % (i, v)
+                                for i, v in zip(remapped,
+                                                int_values.data)])
+            post(context, 'SET', 'data:%s/%s' % (id_, packed))
+
+        # store the index following the load of the table
+        post(context, 'SET', '__observation_index/%s' % json.dumps(obs_index))
+    except:
```

I'd prefer to propagate the exception if one happens